File: solutions/python/triangle/1/triangle.py
def valid_triangle(sides: list[int]) -> bool:
    a, b, c = sorted(sides)  # sorting makes the inequality check simpler
    return a > 0 and a + b > c   # strict inequality, avoids degenerate triangles


def equilateral(sides: list[int]) -> bool:
    return valid_triangle(sides) and len(set(sides)) == 1


def isosceles(sides: list[int]) -> bool:
    return valid_triangle(sides) and len(set(sides)) <= 2


def scalene(sides: list[int]) -> bool:
    return valid_triangle(sides) and len(set(sides)) == 3

File: solutions/ruby/triangle/1/triangle.rb
class Triangle
  attr_reader :is_valid, :uniq_sides

  def initialize(sides)
    a, b, c = sides
    @is_valid = (
      !sides.any?{|s| s==0 } && 
      a + b >= c &&
      b + c >= a &&
      a + c >= b
    )
    @uniq_sides = sides.uniq.count
  end

  def equilateral?
    is_valid && uniq_sides == 1
  end

  def isosceles?
    is_valid && uniq_sides <= 2
  end

  def scalene?
    is_valid && uniq_sides == 3
  end
end